File: xide-standard-plugins/xide-logs-plugin/src/main/java/org/xast/xide/logs_plugin/LogsView.java
package org.xast.xide.logs_plugin;

import org.xast.xide.core.plugin.bottom.BottomPanelView;
import org.xast.xide.core.utils.LucideIcon;
import org.xast.xide.ui.utils.XideStyle;

import java.awt.*;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.swing.*;

public class LogsView extends BottomPanelView {
    private static final int MAX_LINES = 1000;
    private static final int FLUSH_INTERVAL_MS = 500;
    
    private DefaultListModel<LogLine> listModel;
    private JList<LogLine> list;
    private final List<LogLine> allLines = new ArrayList<>();
    private final List<LogLine> pendingLines = new ArrayList<>();
    private final Object pendingLock = new Object();
    private final AtomicBoolean flushRequestPosted = new AtomicBoolean(false);
    private Timer flushTimer;
    private Icon stdOutIcon;
    private Icon stdErrIcon;
    private boolean flushScheduled;
    private boolean showStdOut = true;
    private boolean showStdErr = true;

    public LogsView() {
        setLayout(new BorderLayout());

        XideStyle style = XideStyle.getCurrent();
        Color bgColor = UIManager.getColor("TextArea.background");
        Color fgColor = UIManager.getColor("Label.foreground");

        stdOutIcon = LucideIcon.INFO.icon(16, fgColor);
        stdErrIcon = LucideIcon.CIRCLE_X.icon(16, Color.RED);

        // List
        listModel = new DefaultListModel<>();
        list = new JList<>(listModel);
        list.setBackground(bgColor);
        list.setCellRenderer((listComp, value, index, isSelected, cellHasFocus) -> {
            JLabel label = new JLabel(stripAnsi(value.text()));

            label.setFont(style.uiFont().deriveFont(16f));
            label.setBackground(bgColor);
            label.setBorder(BorderFactory.createEmptyBorder(6, 6, 6, 6));
            label.setIcon(value.isError() ? stdErrIcon : stdOutIcon);
            label.setIconTextGap(8);

            if (isSelected)
                label.setBackground(listComp.getSelectionBackground());

            label.setOpaque(true);

            if (value.isError())
                label.setForeground(Color.RED);

            return label;
        });
        add(
            new JScrollPane(list) {{
                setBorder(BorderFactory.createMatteBorder(1, 0, 0, 0, new Color(0x424242)));
            }}, 
            BorderLayout.CENTER
        );

        flushTimer = new Timer(FLUSH_INTERVAL_MS, e -> flushPendingLines());
        flushTimer.setRepeats(false);

        // Top bar
        Box topBar = Box.createHorizontalBox();
        topBar.setBorder(BorderFactory.createEmptyBorder(4, 0, 4, 0));
        topBar.add(new JCheckBox("StdOut") {{
            setSelected(showStdOut);
            setFont(style.uiFont());
            addActionListener(e -> {
                showStdOut = isSelected();
                rebuildVisibleLines();
            });
        }});
        topBar.add(Box.createRigidArea(new Dimension(8,0)));
        topBar.add(new JCheckBox("StdErr") {{
            setSelected(showStdErr);
            setFont(style.uiFont());
            addActionListener(e -> {
                showStdErr = isSelected();
                rebuildVisibleLines();
            });
        }});
        topBar.add(Box.createRigidArea(new Dimension(8,0)));
        topBar.add(new JButton("Clear") {{
            setForeground(Color.LIGHT_GRAY);
            setFont(style.uiFont());
            addActionListener(e -> {
                synchronized (pendingLock) {
                    pendingLines.clear();
                }
                allLines.clear();
                listModel.clear();
            });
        }});
        add(topBar, BorderLayout.NORTH);

        // Print stream
        System.setOut(new PrintStream(
            new ConsoleRedirectStream(this, false),
            true,
            StandardCharsets.UTF_8
        ));
        System.setErr(new PrintStream(
            new ConsoleRedirectStream(this, true),
            true,
            StandardCharsets.UTF_8
        ));
    }

    public void addLine(String text, boolean error) {
        synchronized (pendingLock) {
            pendingLines.add(new LogLine(text, error));
        }

        requestFlush();
    }

    private void requestFlush() {
        if (!flushRequestPosted.compareAndSet(false, true))
            return;

        SwingUtilities.invokeLater(() -> {
            flushRequestPosted.set(false);

            if (!flushScheduled) {
                flushScheduled = true;
                flushTimer.restart();
            }
        });
    }

    private void flushPendingLines() {
        flushScheduled = false;

        List<LogLine> batch;
        synchronized (pendingLock) {
            if (pendingLines.isEmpty())
                return;

            batch = new ArrayList<>(pendingLines);
            pendingLines.clear();
        }

        allLines.addAll(batch);

        int overflow = allLines.size() - MAX_LINES;
        boolean trimmed = overflow > 0;
        if (trimmed)
            allLines.subList(0, overflow).clear();

        if (trimmed)
            rebuildVisibleLines();
        else
            appendVisibleLines(batch);

        synchronized (pendingLock) {
            if (!pendingLines.isEmpty()) {
                flushScheduled = true;
                flushTimer.restart();
            }
        }
    }

    private void appendVisibleLines(List<LogLine> lines) {
        for (LogLine line : lines) {
            if (shouldShow(line))
                listModel.addElement(line);
        }

        if (!listModel.isEmpty())
            list.ensureIndexIsVisible(listModel.size() - 1);
    }

    private void rebuildVisibleLines() {
        listModel.clear();

        for (LogLine line : allLines) {
            if (shouldShow(line))
                listModel.addElement(line);
        }

        if (!listModel.isEmpty())
            list.ensureIndexIsVisible(listModel.size() - 1);
    }

    private boolean shouldShow(LogLine line) {
        return line.isError() ? showStdErr : showStdOut;
    }

    private String stripAnsi(String text) {
        return text.replaceAll("\u001B\\[[;\\d]*m", "");
    }
}

File: xide-standard-plugins/xide-logs-plugin/src/main/kotlin/org/xast/xide/logs_plugin/LogsView.kt
package org.xast.xide.logs_plugin

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import kotlinx.coroutines.withContext
import kotlin.time.Duration.Companion.milliseconds

import org.xast.xide.core.plugin.bottom.BottomPanelView
import org.xast.xide.core.utils.LucideIcon
import org.xast.xide.ui.utils.XideStyle

import java.awt.*
import java.io.PrintStream
import java.nio.charset.StandardCharsets

import javax.swing.*

class LogsView : BottomPanelView() {
    companion object {
        const val MAX_LINES = 1000
        const val FLUSH_INTERVAL_MS = 500
        private val ANSI_REGEX = Regex("\u001B\\[[\\d]*m")
    }
    
    private val listModel = DefaultListModel<LogLine>()
    private val list = JList(listModel)
    private val allLines = mutableListOf<LogLine>()
    private val scope = CoroutineScope(Dispatchers.Default)
    private val channel = Channel<LogLine>(Channel.UNLIMITED)
    private val stdOutIcon = LucideIcon.INFO.icon(16, UIManager.getColor("Label.foreground"))
    private val stdErrIcon = LucideIcon.CIRCLE_X.icon(16, Color.RED)
    private var showStdOut = true
    private var showStdErr = true

    init {
        layout = BorderLayout()

        val style = XideStyle.getCurrent()
        val bgColor = UIManager.getColor("TextArea.background")
        val fgColor = UIManager.getColor("Label.foreground")

        // List
        list.background = bgColor
        list.setCellRenderer { listComp, value, index, isSelected, cellHasFocus ->
            val label = JLabel(value.text.stripAnsi())

            label.font = style.uiFont()
            label.background = bgColor
            label.border = BorderFactory.createEmptyBorder(6, 6, 6, 6)
            label.icon = if (value.isError) stdErrIcon else stdOutIcon
            label.iconTextGap = 8

            if (isSelected)
                label.background = listComp.selectionBackground

            label.isOpaque = true

            if (value.isError)
                label.foreground = Color.RED

            label
        }

        add(
            JScrollPane(list).apply {
                border = BorderFactory.createMatteBorder(1, 0, 0, 0, Color(0x424242))
            }, 
            BorderLayout.CENTER
        )

        // Top bar
        var topBar = Box.createHorizontalBox()
        topBar.border = BorderFactory.createEmptyBorder(4, 0, 4, 0)
        topBar.add(JCheckBox("StdOut").apply {
            isSelected = showStdOut
            font = style.uiFont()
            addActionListener {
                showStdOut = isSelected
                rebuildVisibleLines()
            }
        })
        topBar.add(Box.createRigidArea(Dimension(8,0)))
        topBar.add(JCheckBox("StdErr").apply{
            isSelected = showStdErr
            font = style.uiFont()
            addActionListener {
                showStdErr = isSelected
                rebuildVisibleLines()
            }
        })
        topBar.add(Box.createRigidArea(Dimension(8,0)))
        topBar.add(JButton("Clear").apply {
            setForeground(Color.LIGHT_GRAY)
            setFont(style.uiFont())
            addActionListener {
                while (true) {
                    if (channel.tryReceive().getOrNull() == null)
                        break
                }
                allLines.clear()
                listModel.clear()
            }
        })
        add(topBar, BorderLayout.NORTH)

        scope.launch {
            val buffer = mutableListOf<LogLine>()

            while (true) {
                val first = channel.receive()
                buffer.add(first)

                delay(FLUSH_INTERVAL_MS.milliseconds)

                while (true) {
                    val next = channel.tryReceive().getOrNull() ?: break
                    buffer.add(next)
                }

                val batch = buffer.toList()
                buffer.clear()

                withContext(Dispatchers.Swing) {
                    flushBatch(batch)
                }
            }
        }

        // Print stream
        System.setOut(PrintStream(
            ConsoleRedirectStream(this, false),
            true,
            StandardCharsets.UTF_8
        ))
        System.setErr(PrintStream(
            ConsoleRedirectStream(this, true),
            true,
            StandardCharsets.UTF_8
        ))
    }

    fun addLine(text: String, error: Boolean) {
        channel.trySend(LogLine(text, error))
    }

    private fun flushBatch(batch: List<LogLine>) {
        allLines.addAll(batch)

        val overflow = allLines.size - MAX_LINES
        val trimmed = overflow > 0
        if (trimmed)
            allLines.subList(0, overflow).clear()

        if (trimmed)
            rebuildVisibleLines()
        else
            appendVisibleLines(batch)
    }

    private fun appendVisibleLines(lines: List<LogLine>) {
        for (line in lines) {
            if (shouldShow(line))
                listModel.addElement(line)
        }

        if (listModel.size > 0)
            list.ensureIndexIsVisible(listModel.size - 1)
    }

    private fun rebuildVisibleLines() {
        listModel.clear()

        for (line in allLines) {
            if (shouldShow(line))
                listModel.addElement(line)
        }

        if (listModel.size > 0)
            list.ensureIndexIsVisible(listModel.size - 1)
    }

    private fun shouldShow(line: LogLine): Boolean =
        if (line.isError) showStdErr else showStdOut

    private fun String.stripAnsi(): String =
        replace(ANSI_REGEX, "")
}
